### PR TITLE
NukeOps game rule tweaks

### PR DESCRIPTION
**Pull request by TekuNut**
Enables nuke-ops game rule to be activated mid-round. 

- Spawn ghost role spawn-points for nuclear operatives when started mid-round. This uses the same formula as is used on round start to determine the number of spawn points.
- Fix edge-case where round would end if a sole nuke operator is in critical condition.
- Move hard-coded values into `NukeopsRuleConfigPrototype`. Can be overridden by providing a prototype name via CVar `nukeops.config_prototype`.

### CODE REVIEW

**Comment by TekuNut** [comment]
Some questions:
- If a player is made into a lone operative and the game rule is currently running or later added, should they be considered as a part of the nuke operative team? Or kept as a separate entity?
- If the number of ghosts able to be a nuke op is too low, should the game rule abort when adding mid-round?

**Comment by keronshb** [comment]
> Some questions:
> 
>     * If a player is made into a lone operative and the game rule is currently running or later added, should they be considered as a part of the nuke operative team? Or kept as a separate entity?
> 
>     * If the number of ghosts able to be a nuke op is too low, should the game rule abort when adding mid-round?

I'm not a maintainer but this is how I've seen it work in Dynamic in SS13:

Lone Ops are well, considered to be working alone even though they can be syndicate sided (with Nukies, Traitors, Radio Op, etc). I've seen them team up with Nukies since their goal is the same however the difference is that the Lone Op gets the code to set off the stationary Station Nuke where as Nuclear Operatives gets the code to set off the Nuclear Fission Device that's on their ship. Same end goal, but technically separate teams.

IIRC with Dynamic, if not enough ghosts ready up the points are refunded and a different event fires off.

**Comment by TekuNut** [comment]
Thanks clearing up the Lone Nukie situation :+1: .

Regarding the dynamic round stuff, I think a change/addition to ghost roles to have the game rule ask a system for some number of ghosts interested in a role/group of roles would be a good step towards this. (I'm thinking just adding this to GhostRoleSystem would do). Once the limit of players is reached (or it waits the full time allowance and picks randomly) this system calls the game rule back with the players interested, or a failure result if not enough players signed up.

**Comment by Peptide90** [comment]
> Thanks clearing up the Lone Nukie situation 👍 .
> 
> Regarding the dynamic round stuff, I think a change/addition to ghost roles to have the game rule ask a system for some number of ghosts interested in a role/group of roles would be a good step towards this. (I'm thinking just adding this to GhostRoleSystem would do). Once the limit of players is reached (or it waits the full time allowance and picks randomly) this system calls the game rule back with the players interested, or a failure result if not enough players signed up.

#8928

**Comment by moonheart08** [comment]
You should be using a custom GameRuleConfiguration, not a brand new prototype.

File: Content.Server/GameTicking/Rules/Components/NukeOperativeSpawnerComponent.cs
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/Components/NukeOperativeSpawnerComponent.cs:18]: Should use Access to ensure only nukeopsrulesystem sets these.

File: Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs:338]: Null is already checked above at line 326.

File: Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs:391]: Should move these cvars onto the prototype.

**Review by metalgearsloth — DISMISSED** [review]

File: Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs
Comment (TekuNut) [reply Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs]: Done. I've replaced it with an entity query. In theory, other entities can have the NukeOperativeComponent added by admins and they will be counted as being on the nuke operative team.

Some questions:
- Should the round end if there are nuke op ghost roles still available?
- If other entities did get the NukeOperativeComponent added, should they be part of the killed operative check? I'm considering adding a bool datafield to indicate this.

Comment (vulppine) [reply Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs]: for your first question, i feel like it should if the ghost role operatives aren't able to reasonably get to the station (i.e., their shuttle is not on the moon base map, and they're stuck on the nukie base as a result if all other operatives are dead)

for consideration, though: leaving the round running *may* come with a chance that the crew would attempt to invade the nuclear operative base through the operative shuttle *if* they have control over it, meaning that the operatives can try to take back their ship and continue the mission, but that's all dependent on the players and it may drag out a round for too long

Comment (TekuNut) [reply Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs]: For now, I'll have it check for nuke operatives being alive on the same map that the shuttle is present in. This does introduce a problem, where the round will end if the shuttle is hijacked and taken to a map that isn't the outpost or the station and the check is triggered. 

`CheckRoundShouldEnd` should also check if operatives are present on the target station. I'm not sure if I should leave that to be implemented in #10220 , or take some code from that PR and implement the logic in this branch?

**Comment by metalgearsloth** [comment]
Get a green tick and it's good for merge

File: Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs
Comment (vulppine) [reply Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs]: i can implement this behavior in #10220 once this is merged in, since the scope of that focuses all on nuclear operative win conditions

**Review by mirrorcult — APPROVED** [review]

**Pull request merged**
